examples/src/test/java/io/grpc/examples/header/HeaderServerInterceptorTest.java

```diff
@@ -0,0 +1,136 @@
+import io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import io.grpc.ManagedChannel;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Server;
+import io.grpc.ServerInterceptors;
+import io.grpc.examples.helloworld.GreeterGrpc;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterBlockingStub;
+import io.grpc.examples.helloworld.GreeterGrpc.GreeterImplBase;
+import io.grpc.examples.helloworld.HelloReply;
+import io.grpc.examples.helloworld.HelloRequest;
+import io.grpc.inprocess.InProcessChannelBuilder;
+import io.grpc.inprocess.InProcessServerBuilder;
+import io.grpc.stub.StreamObserver;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.ArgumentCaptor;
+
+/**
+ * Unit tests for {@link HeaderClientInterceptor}.
+ * For demonstrating how to write gRPC unit test only.
+ * Not intended to provide a high code coverage or to test every major usecase.
+ *
+ * <p>For basic unit test examples see {@link io.grpc.examples.helloworld.HelloWorldClientTest} and
+ * {@link io.grpc.examples.helloworld.HelloWorldServerTest}.
+ */
+@RunWith(JUnit4.class)
+public class HeaderServerInterceptorTest {
+  private Server fakeServer;
+  private ManagedChannel inProcessChannel;
+
+  @Before
+  public void setUp() throws Exception {
+    String uniqueServerName = "fake server for " + getClass();
+    GreeterImplBase greeterImplBase =
+        new GreeterImplBase() {
+          @Override
+          public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) {
+            responseObserver.onNext(HelloReply.getDefaultInstance());
+            responseObserver.onCompleted();
+          }
+        };
+    fakeServer = InProcessServerBuilder.forName(uniqueServerName)
+        .addService(ServerInterceptors.intercept(greeterImplBase, new HeaderServerInterceptor()))
+        .directExecutor()
+        .build()
+        .start();
+    inProcessChannel = InProcessChannelBuilder.forName(uniqueServerName).directExecutor().build();
+  }
+
+  @After
+  public void tearDown() {
+    inProcessChannel.shutdownNow();
+    fakeServer.shutdownNow();
+  }
+
+  @Test
+  public void serverHeaderDeliveredToClient() {
+    class SpyingClientInterceptor implements ClientInterceptor {
+      ClientCall.Listener<?> spyListener;
+
+      @Override
+      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+          MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+        return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+          @Override
+          public void start(Listener<RespT> responseListener, Metadata headers) {
+            spyListener = spy(responseListener);
+            @SuppressWarnings("unchecked") // absolutely casting to the correct type
+            Listener<RespT> castedSpyListener = (Listener<RespT>) spyListener;
```

Comment:
This cast is unnecessary and is caused by assigning to the `<?>` variable (I don't care whether you split the assignment into two lines or use one):
```java
spyListener = responseListener = spy(responseListener);
super.start(responseListener, headers);
```